src/TodayEventMonitor.vala

```diff
@@ -0,0 +1,181 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+public class Calendar.TodayEventMonitor : GLib.Object {
+    private Gee.HashMultiMap<ECal.Component, string> event_uids;
+
+    construct {
+        load_today_events ();
+        Timeout.add_seconds (86400, () => {
+            event_uids.clear ();
+            load_today_events ();
+            return true;
+        });
+    }
+
+    private void load_today_events () {
+        event_uids = new Gee.HashMultiMap<ECal.Component, string> ();
+        var model = new Calendar.EventStore ();
+        model.events_added.connect (on_events_added);
+        model.events_updated.connect (on_events_updated);
+        model.events_removed.connect (on_events_removed);
+        model.month_start = Calendar.Util.datetime_get_start_of_month (new DateTime.now_local ());
+    }
+
+    private void on_events_added (E.Source source, Gee.Collection<ECal.Component> events) {
+        var extension = (E.SourceAlarms)source.get_extension (E.SOURCE_EXTENSION_ALARMS);
+        if (extension.get_include_me () == false) {
+            return;
+        }
+
+        Idle.add ( () => {
+            foreach (var event in events)
+                add_event (source, event);
+
+            return false;
+        });
+    }
+
+    private void on_events_updated (E.Source source, Gee.Collection<ECal.Component> events) {
+        Idle.add ( () => {
+            foreach (var event in events)
+                update_event (source, event);
+
+            return false;
+        });
+    }
+
+    private void on_events_removed (E.Source source, Gee.Collection<ECal.Component> events) {
+        Idle.add ( () => {
+            foreach (var event in events)
+                remove_event (source, event);
+
+            return false;
+        });
+    }
+
+    private void add_event (E.Source source, ECal.Component event) {
+        unowned ICal.Component comp = event.get_icalcomponent ();
+        debug ("Event [%s, %s, %s]".printf (comp.get_summary (), source.dup_display_name (), comp.get_uid ()));
+        foreach (var alarm_uid in event.get_alarm_uids ()) {
+            ECal.ComponentAlarm e_alarm = event.get_alarm (alarm_uid);
+            ECal.ComponentAlarmAction action;
+
+#if E_CAL_2_0
+            action = e_alarm.get_action ();
+#else
+            e_alarm.get_action (out action);
+#endif
+            switch (action) {
+                case (ECal.ComponentAlarmAction.DISPLAY):
```

Comment:
This probably makes more sense as an if statement? There's currently only one case, so it would make sense that way as an if/else. But we can leave it as is to make future action types easier to add, if you want.

Reply:
Yeah, an if makes more sense here.